src/runtime_src/driver/xclng/drm/xocl/subdev/nifd.c

```diff
@@ -0,0 +1,680 @@
+/*
+ * A GEM style device manager for PCIe nifd_based OpenCL accelerators.
+ *
+ * Copyright (C) 2016-2018 Xilinx, Inc. All rights reserved.
+ *
+ * Authors:
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#define pr_fmt(fmt) KBUILD_MODNAME ":%s: " fmt, __func__
+
+#include <linux/types.h>
+#include <linux/uaccess.h>
+#include <linux/vmalloc.h>
+#include <linux/slab.h>
+#include <linux/cdev.h>
+#include <linux/fs.h>
+#include <linux/device.h>
+#include <linux/io.h>
+#include <linux/ioctl.h>
+
+#include "../xocl_drv.h"
+#include "xclfeatures.h"
+
+#define NIFD_DEV_NAME "nifd" SUBDEV_SUFFIX
+#define SUPPORTED_NIFD_IP_VERSION 1
+#define SUPPORTED_DRIVER_VERSION 1
+#define MINOR_NAME_MASK 0xffffffff
+
+enum NIFD_register_offset
+{
+    NIFD_START_APP = 0x0,
+    NIFD_STOP_APP = 0x4,
+    NIFD_CLEAR = 0x8,
+    NIFD_CLEAR_CFG = 0xc,
+    NIFD_CLEAR_BREAKPOINT = 0x10,
+    NIFD_CLK_MODES = 0x14,
+    NIFD_START_READBACK = 0x18,
+    NIFD_CLOCK_COUNT = 0x1c,
+    NIFD_CONFIG_DATA = 0x20,
+    NIFD_BREAKPOINT_CONDITION = 0x24,
+    NIFD_STATUS = 0x28,
+    NIFD_CLOCK_CNT = 0x2c,
+    NIFD_READBACK_DATA = 0x30,
+    NIFD_READBACK_DATA_WORD_CNT = 0x34,
+    NIFD_CONFIG_DATA_M2 = 0x38,
+    NIFD_CLEAR_CFG_M2 = 0x3c
+};
+
+enum NIFD_COMMAND_SEQUENCES
+{
+    NIFD_ACQUIRE_CU = 0,
+    NIFD_RELEASE_CU = 1,
+    NIFD_QUERY_CU = 2,
+    NIFD_READBACK_VARIABLE = 3,
+    NIFD_SWITCH_ICAP_TO_NIFD = 4,
+    NIFD_SWITCH_ICAP_TO_PR = 5,
+    NIFD_ADD_BREAKPOINTS = 6,
+    NIFD_REMOVE_BREAKPOINTS = 7,
+    NIFD_CHECK_STATUS = 8,
+    NIFD_QUERY_XCLBIN = 9,
+    NIFD_STOP_CONTROLLED_CLOCK = 10,
+    NIFD_START_CONTROLLED_CLOCK = 11,
+    NIFD_SWITCH_CLOCK_MODE = 12
+};
+
+struct xocl_nifd
+{
+    void *__iomem nifd_base;
+    void *__iomem icap_base;
+    unsigned int instance;
+    struct cdev sys_cdev;
+    struct device *sys_device;
+};
+
+static dev_t nifd_dev;
+
+struct xocl_nifd *nifd_global;

```

Comment:
How do we support multiple boards with a system-wide global pointer?

Reply:
so does nifd_valid blow.

Reply:
@maxzhen in the new push, I changed it to pass through ioctl instead of using a global variable.

Reply:
@houlz0507 in the new push, I changed it to do early return in the probe and got rid of the nifd_valid global flag.

Reply:
It is clean. Thanks